Feature/state storage pdo expiry

pmeulen opened the pull request:
Strictly adhere to the expiry when using the pdo state storage. This brings the behavior in line with the of the memcache implementation and this means that the frequency at which the GC runs can be much, much lower.

Todo:
- Move GC run (cleanExpired) to setValue(). This will make the number of times that the GC runs more predictable
- Make the probability that the GC runs configurable

**Edit Michiel**: 
I opened this PR based on your work: #25 merging that into this branch should get us another step in the right direction.

pmeulen committed:
Handle expiry the memcache way

Never return a expired value
An expiry of 0 means never expire

pmeulen committed:
Add tests fors StateStorage for pdo and file types

MKodde committed:
Install Mockery for testing purposes

MKodde committed:
Refactor PDO StateStorage

Goal is to add an additional configuration parameter (the probability
that the expired state is cleaned up) But other improvements have been
applied that where encountered and had little or no impact but
significantly improved the class.

1. Made the table a specific parameter instead of grabbin it from the
   'options' array
2. Did the same for the 'cleanupProbability' configuration as per [1]
   (task 2)
3. Moved the cleanup of expiry data to the 'setValue' method as per [1]
   (task 1)

[1] https://www.pivotaltracker.com/story/show/181553917

MKodde committed:
Integrate StateStorage PDO changes in factory

The factory now has to unpack the options, and verify the correct
options are present in the configuration. If they are not, a Runtime
exception is thrown. This can later be catched, or not if we want this
to halt the execution of the code path.

MKodde reviewed (approved):
Thanks for providing this starting point! Some constructions where slightly changed in #25 to make them fit into the test behaviors I added.

MKodde committed:
Change probability calculating algorithm

Instead of testing on a range of 0..1000. We now set a probability based
on a percentage expressed in a floating point value between zero and
one. Where 0 is 0% and 1 expresses 100%.

This makes configuration more predictable.

Thanks for your suggestion and code contribution @pmeulen:
https://github.com/SURFnet/tiqr-server-libphp/pull/25/files#r837431802

pmeulen committed:
Merge pull request #25 from SURFnet/feature/move-expiry-period-and-make-configurable

Move expiry action and make probability of triggering it configurable
